python/tester/upload.py

```diff
@@ -84,34 +153,222 @@ def main():
 
+    last_group = get_last_group(storage_client, instrument, date)
+    _log.info(f"Last group {last_group}")
+
+    src_bucket = storage_client.bucket("rubin-prompt-proto-unobserved")
+    raw_pool = get_samples(src_bucket, instrument)
+
+    if raw_pool:
+        _log.info(f"Observing real raw files from {instrument}.")
+        for i, true_group in enumerate(itertools.islice(itertools.cycle(raw_pool), n_groups)):
+            group = last_group + i + 1
+            _log.debug(f"Processing group {group} from unobserved {true_group}...")
+            # snap_dict maps snap_id to {visit: blob}
+            snap_dict = {}
+            for snap_id, old_visits in raw_pool[true_group].items():
+                snap_dict[snap_id] = {splice_group(true_visit, group): blob
+                                      for true_visit, blob in old_visits.items()}
+            visit_infos = {info for det_dict in snap_dict.values() for info in det_dict}
+
+            # TODO: may be cleaner to use a functor object than to depend on
+            # closures for the bucket and data.
+            def upload_from_pool(visit, snap_id):
+                src_blob = snap_dict[snap_id][visit]
+                exposure_id = make_exposure_id(visit.instrument, visit.group, snap_id)
+                fname = raw_path(visit.instrument, visit.detector, visit.group, snap_id,
+                                 exposure_id, visit.filter)
+                src_bucket.copy_blob(src_blob, dest_bucket, new_name=fname)
+            process_group(publisher, visit_infos, upload_from_pool)
+            _log.info("Slewing to next group")
+            time.sleep(SLEW_INTERVAL)
+    else:
+        _log.info(f"No raw files found for {instrument}, generating dummy files instead.")
+        for i in range(n_groups):
+            group = last_group + i + 1
+            visit_infos = make_random_visits(instrument, group)
+
+            # TODO: may be cleaner to use a functor object than to depend on
+            # closures for the bucket and data.
+            def upload_dummy(visit, snap_id):
+                exposure_id = make_exposure_id(visit.instrument, visit.group, snap_id)
+                fname = raw_path(visit.instrument, visit.detector, visit.group, snap_id,
+                                 exposure_id, visit.filter)
+                dest_bucket.blob(fname).upload_from_string("Test")
+            process_group(publisher, visit_infos, upload_dummy)
+            _log.info("Slewing to next group")
+            time.sleep(SLEW_INTERVAL)
+
+
+def get_last_group(storage_client, instrument, date):
+    """Identify a group number that will not collide with any previous groups.
+
+    Parameters
+    ----------
+    storage_client : `google.cloud.storage.Client`
+        A Google Cloud Storage object pointing to the active project.
+    instrument : `str`
+        The short name of the active instrument.
+    date : `str`
+        The current date in YYYYMMDD format.
```

Comment:
Not a datetime or astropy date?

Reply:
No. This is part of the original group generation code, and I don't want to touch it because then I'd need to come up with a different scheme for avoiding collisions between uploaded files.